topdown/jsonschema.go

```
@@ -0,0 +1,139 @@
+	case ast.Object:
+		var data []byte
+		var asJSON interface{}
+		asJSON, err = ast.JSON(value)
+		if err != nil {
+			break
+		}
+		data, err = json.Marshal(asJSON)
+		loader = gojsonschema.NewStringLoader(string(data))
+	default:
+		return nil, errors.New("wrong type, expected string or object")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return loader, nil
+}
+
+// builtinJSONSchemaIsValid accepts 1 argument which can be string or object and checks if it is valid JSON schema.
+func builtinJSONSchemaIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	loader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return iter(ast.BooleanTerm(false))
+	}
+
+	if _, err = gojsonschema.NewSchema(loader); err != nil {
+		return iter(ast.BooleanTerm(false))
+	}
+
+	return iter(ast.BooleanTerm(true))
+}
+
+// builtinJSONSchemaValidate accepts 1 argument which can be string or object and checks if it is valid JSON schema.
+// Returns string in case of error or empty string otherwise.
+func builtinJSONSchemaValidate(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	loader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return iter(ast.StringTerm("jsonschema: " + err.Error()))
+	}
+
+	if _, err = gojsonschema.NewSchema(loader); err != nil {
+		return iter(ast.StringTerm("jsonschema: " + err.Error()))
+	}
+
+	return iter(ast.StringTerm(""))
+}
+
+// builtinJSONMatchSchema accepts 2 arguments both can be string or object and verifies if the document matches the JSON schema.
+// Returns an array of errors or empty array.
+// Returns an empty array if no errors are found.
+// In case of internal error returns empty array.
+func builtinJSONMatchSchema(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	documentLoader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return iter(ast.ArrayTerm())
```

I think this should return the error here and below. Successful validation can have an empty array but errors should not.